docs/tutorial/02-making-improvements.md

```diff
@@ -0,0 +1,269 @@
+
+```bash
+$ stellar contract invoke --id guess_the_number --source me --network local -- help
+Commands:
+  reset    Update the number. Only callable by admin.
+  guess    Guess a number between 1 and 10
+  upgrade  Upgrade the contract to new wasm. Only callable by admin.
+  help     Print this message or the help of the given subcommand(s)
+```
+
+It would error if you tried to invoke it:
+
+```bash
+$ stellar contract invoke --id guess_the_number --source me --network local -- set_random_number
+error: unrecognized subcommand 'set_random_number'
+```
+
+### Wait, So Anyone Can Call Reset?
+
+Nope! Just because we made it public, we still require authentication so only admins can call it. Rust's idea of public vs private handles "where" the functions can be called. You still need to handle "who" calls it. That's why we set the contract admin in it's constructor method and check it with `Self::require_admin(env);`. Let's test it out and create a non-admin identity to see how it fails:
+
+```bash
+$ stellar keys generate bob --network local --fund
+✅ Key saved with alias bob in ".config/stellar/identity/bob.toml"
+✅ Account bob funded on "Standalone Network ; February 2017"
+
+$ stellar contract invoke --id guess_the_number --source bob --network local -- reset
+❌ error: Missing signing key for account GDAQWVA6REGN47BBCFY6SGQ4YTIGMDZZFHDOVUZXMVRAAT6OEZGCACGH
+```
+
+Alice is the admin, Bob is just a regular user. She can call `reset`, he gets an error.
+
+## Step 4: ⚠️ Improve Error Handling
+
+Speaking of errors, we should provide some helpful messages instead of just breaking when things go wrong. Let's make the `guess` function more robust by replacing `unwrap()` with `expect()`:
+
+```rust
+/// Guess a number between 1 and 10
+pub fn guess(env: &Env, a_number: u64) -> bool {
+    let stored_number = env.storage()
+        .instance()
+        .get::<_, u64>(&THE_NUMBER)
+        .expect("No number has been set");
+
+    a_number == stored_number
+}
+```
+
+### Understanding `expect()` vs `unwrap()`
+
+Because [Rust distinguishes between two types of errors](https://doc.rust-lang.org/book/ch09-00-error-handling.html), there's two different ways to handle them. Some errors are from bugs, and obviously we want to avoid these as much as possible. But some errors are expected. And we can recover from them without crashing. Both of these methods are shortcuts to check for values and error, but one gives you the ability to explain _why_:
+
+- `unwrap()`: Crashes with a generic error message
+- `expect()`: Crashes with your custom error message
+
+Both will panic if the value is `None`, but `expect()` gives better debugging info. In our case, this should never happen since we now set a number in the constructor, but it's good defensive programming. For example, we were already using it in our `require_admin` method for authentication.
```

Comment:
Hmm I need to check but I thought that message provided by expect are not surfaced when calling the contract. If so, perhaps best to leave this until we introduce errors or introduce them now?